### ISSUE TITLE
Bug: all_gather_cat silently corrupts metric shapes and values on single-device runs

### ISSUE THREAD

**Issue by RajdeepKushwaha5** [mllam/neural-lam#421]
## Description

`ARModel.all_gather_cat` unconditionally calls `.flatten(0, 1)` on the result of `self.all_gather()`. On multi-GPU (DDP), `all_gather` prepends a new dimension `(K, d1, d2, ...)`, so flattening dims 0 and 1 correctly produces `(K*d1, d2, ...)`. However, on a single device, `all_gather` returns the tensor **unchanged** — no new leading dimension is added. This means `flatten(0, 1)` **merges the first two existing dimensions**, silently producing wrong tensor shapes and therefore wrong metric values.

## Current behavior

On single GPU/CPU, in `aggregate_and_plot_metrics`:

```
Input to all_gather_cat:  (N_eval, pred_steps, d_f)
After all_gather:         (N_eval, pred_steps, d_f)  ← no new dim on single device
After flatten(0, 1):      (N_eval * pred_steps, d_f)  ← WRONG: merged batch and time dims
After mean(dim=0):         (d_f,)                      ← WRONG: should be (pred_steps, d_f)
```

This means:
- **Validation/test metrics are silently averaged across both samples AND time steps**, instead of just across samples
- Downstream `plot_error_map` receives a 1D `(d_f,)` tensor instead of 2D `(pred_steps, d_f)`, causing a crash (partially patched by #193 in `vis.py`, but the values remain wrong)
- `on_test_epoch_end` spatial loss maps similarly have corrupted shapes: `(N_test * N_log, num_grid_nodes)` instead of `(N_test, N_log, num_grid_nodes)`, giving nonsensical spatial loss averages

## Expected behavior

On single device, `all_gather_cat` should return the tensor unchanged since there is nothing to gather/concatenate from other ranks.

## Reproduction

```python
import torch
import pytorch_lightning as pl

# Simulate single-device all_gather behavior:
tensor = torch.randn(4, 3, 5)  # (N_eval, pred_steps, d_f)
gathered = tensor  # SingleDeviceStrategy.all_gather returns tensor unchanged
result = gathered.flatten(0, 1)  # BUG: (12, 5) instead of (4, 3, 5)
averaged = result.mean(dim=0)   # (5,) instead of (3, 5)

print(f"Input: {tensor.shape}")
print(f"After flatten(0,1): {result.shape}")     # (12, 5) ← WRONG
print(f"After mean(dim=0): {averaged.shape}")     # (5,)   ← WRONG
```

## Root cause

[`ar_model.py` line 331](https://github.com/mllam/neural-lam/blob/main/neural_lam/models/ar_model.py#L331):
```python
def all_gather_cat(self, tensor_to_gather):
    return self.all_gather(tensor_to_gather).flatten(0, 1)
```

PyTorch Lightning's `SingleDeviceStrategy.all_gather` simply returns the tensor unchanged ([source](https://github.com/Lightning-AI/pytorch-lightning/blob/master/src/lightning/pytorch/strategies/single_device.py)):
```python
def all_gather(self, tensor, group=None, sync_grads=False):
    return tensor
```

## Proposed fix

Add a dimension guard before flattening:

```python
def all_gather_cat(self, tensor_to_gather):
    gathered = self.all_gather(tensor_to_gather)
    if gathered.dim() > tensor_to_gather.dim():
        return gathered.flatten(0, 1)
    return gathered
```

## Affected code paths

- `aggregate_and_plot_metrics` (line 647): validation and test metric aggregation
- `on_test_epoch_end` (line 700): spatial loss map aggregation

## Related

- #192 / PR #193 — patches the downstream crash in `vis.py` but does **not** fix the root cause (metrics remain silently wrong)
- PR #208 — includes this fix bundled in a large refactor, but that PR is not yet merged

## Impact

**High** — Any single-GPU training or evaluation silently produces wrong validation/test metrics. Users get metrics that are averaged across both time steps and samples, giving misleadingly low error values. The test suite itself acknowledges this with the comment:
```python
# XXX: `devices` has to be set to 2 otherwise
# neural_lam.models.ar_model.ARModel.aggregate_and_plot_metrics fails
# because it expects to aggregate over multiple devices
devices=2,
```

**Comment by joeloskarsson** [comment]
Is this actually a bug when you run with single device and ddp strategy? Which is the only setting present on main. It seems like me as something we might potentially run into if we switch the trainer strategy from ddp to auto, as discussed in #193

**Comment by RajdeepKushwaha5** [comment]
> Is this actually a bug when you run with single device and ddp strategy? Which is the only setting present on main. It seems like me as something we might potentially run into if we switch the trainer strategy from ddp to auto, as discussed in [#193](https://github.com/mllam/neural-lam/pull/193)

@joeloskarsson You're right — I should have been more precise in the issue description. Thank you for the sanity check.

I went back and verified this end-to-end against the Lightning internals (PL 2.6.1):

With `strategy="ddp"` hardcoded on `main`, even when a user runs on a single CPU/GPU with `--devices auto` (which resolves to 1 device), Lightning still instantiates `DDPStrategy` with `_SubprocessScriptLauncher`. The subprocess calls `setup_distributed()` → `_init_dist_connection()`, which initializes the process group (world_size=1). At that point, `_all_gather_ddp_if_available` sees `_distributed_is_initialized() == True`, runs the real `torch.distributed.nn.functional.all_gather`, and returns `torch.stack(gathered)` — shape `(1, d1, d2, ...)`. So `flatten(0, 1)` correctly produces `(d1, d2, ...)`.

**So no — this is not an active bug on the current `main` production path.** The hardcoded `strategy="ddp"` masks it completely.

That said, two things worth noting:

1. **The test suite on `main`** doesn't set `strategy` (it defaults to `"auto"`) and uses `devices=2` with the [XXX comment](https://github.com/mllam/neural-lam/blob/8141679/tests/test_training.py#L47-L49) explicitly acknowledging that `aggregate_and_plot_metrics` breaks otherwise. If the tests were ever changed to `devices=1` (without also adding `strategy="ddp"`), `SingleDeviceStrategy` would be used and the bug would surface.

2. **PR #193** proposes changing `strategy="ddp"` → `strategy="auto"` in `train_model.py` itself, plus `devices=2 if torch.cuda.device_count() >= 2 else 1` in tests. If/when that lands, single-device users would hit `SingleDeviceStrategy` and the bug becomes active on the production path.

I should have framed this as a **latent issue in `all_gather_cat` that assumes DDP is always active**, rather than claiming it affects current production runs. The fix (a dimension guard) is a no-op when DDP is active — `gathered.dim() > tensor_to_gather.dim()` is always `True` — so it carries zero risk on the current `main`. Its value is in making `all_gather_cat` strategy-agnostic, which prepares for the `strategy="auto"` switch discussed in #193 and unblocks the `devices=2` workaround removal.

I'll update the issue title and description to reflect this more accurately. Let me know if you'd prefer to defer the fix to be paired with the `strategy="auto"` change, or if the standalone PR is acceptable as a forward-looking hardening.

**Comment by joeloskarsson** [comment]
Ok great, thanks for the explanation. This is a good fix to add for robustness then, and I think we will merge this before #193 to avoid the issues.

**Comment by RajdeepKushwaha5** [comment]
> Ok great, thanks for the explanation. This is a good fix to add for robustness then, and I think we will merge this before [#193](https://github.com/mllam/neural-lam/pull/193) to avoid the issues.

Thanks Sir — glad the explanation helped clarify things. Good to hear it'll go in ahead of #193; that should make the strategy switch cleaner when the time comes.
